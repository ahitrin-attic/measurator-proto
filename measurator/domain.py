import datetime


class IO:
    """Custom wrapper for all input/output interactions"""

    def write(self, text: str, *args) -> None:
        raise NotImplementedError

    def read(self) -> str:
        raise NotImplementedError

    def write_file(self, data: iter) -> None:
        raise NotImplementedError

    def read_file(self) -> iter:
        raise NotImplementedError

    def now(self) -> datetime.datetime:
        raise NotImplementedError
