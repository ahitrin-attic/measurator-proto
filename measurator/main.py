import argparse
import csv
import datetime
import time


def migrate():
    pass


def run_main():
    path = _file_path()
    not_yet, succeeds, fails = _read_file(path)
    # evaluate measurements
    now = datetime.datetime.now()
    delayed = list()
    for status, timestamp, text in not_yet:
        evaluate_time = datetime.datetime(*(time.strptime(timestamp, "%Y-%m-%d %H:%M")[:6]))
        if evaluate_time < now:
            print("Time to evaluate: {}\n Is it true? (Yes/No/Delay)".format(text))
            user_input = input().capitalize()
            if user_input.startswith("Y"):
                status = "S"
                succeeds.append((status, timestamp, text))
            elif user_input.startswith("D"):
                print("When to evaluate (YYYY-mm-dd HH:MM):")
                eval_time = input()
                delayed.append(("N", eval_time, text))
            else:
                status = "F"
                fails.append((status, timestamp, text))
    not_yet = delayed
    # print total statistics
    total_done = len(fails) + len(succeeds)
    if total_done > 0:
        percentage = "%d%%" % (float(100 * len(succeeds)) / float(total_done))
    else:
        percentage = "N/A"
    print("Succesful predictions:", percentage, ", not done yet:", len(not_yet))
    # add another prediction when needed
    print("Add another prediction? Y/N")
    user_input = input()
    if user_input.capitalize().startswith("Y"):
        print("Prediction:")
        prediction = input()
        print("When to evaluate (YYYY-mm-dd HH:MM):")
        eval_time = input()
        not_yet.append(("N", eval_time, prediction))
    # overwrite predictions file
    all_rows = list()
    all_rows.extend(fails)
    all_rows.extend(succeeds)
    all_rows.extend(not_yet)
    with open(path, "wt") as f:
        writer = csv.writer(f)
        for row in all_rows:
            writer.writerow(row)


def _read_file(path):
    fails = list()
    succeeds = list()
    not_yet = list()
    try:
        with open(path) as f:
            reader = csv.reader(f)
            for row in reader:
                status = row[0]
                if status == "F":
                    fails.append(row)
                elif status == "S":
                    succeeds.append(row)
                else:
                    not_yet.append(row)
    except FileNotFoundError:
        pass
    return not_yet, succeeds, fails


def _file_path():
    parser = argparse.ArgumentParser()
    parser.add_argument("path")
    args = parser.parse_args()
    return args.path
