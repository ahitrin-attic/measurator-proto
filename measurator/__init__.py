from measurator.main import run_main, migrate
