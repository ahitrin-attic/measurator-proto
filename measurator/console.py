import datetime

from measurator.domain import IO


class ConsoleIO(IO):
    def __init__(self, path) -> None:
        super().__init__()
        self._path = path

    def write(self, text: str, *args):
        print(text, *args)

    def read(self) -> str:
        return input()

    def write_file(self, data: iter) -> None:
        with open(self._path, "w") as f:
            for line in data:
                f.write(line)

    def read_file(self) -> iter:
        try:
            with open(self._path) as f:
                return f.readlines()
        except FileNotFoundError:
            return []

    def now(self) -> datetime.datetime:
        return datetime.datetime.now()
