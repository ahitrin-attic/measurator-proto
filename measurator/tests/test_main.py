import datetime
from typing import List

from approvaltests import verify
from approvaltests.reporters import GenericDiffReporterFactory

from measurator import run_main_
from measurator.domain import IO


class TestableIO(IO):
    def __init__(self, inputs: List[str], timestamps: List[datetime.datetime], file_content: iter):
        self.file_content = file_content
        self.inputs = inputs
        self.timestamps = timestamps
        self.log: List[str] = []

    def write(self, text: str, *args) -> None:
        self.log.append(f" {text}{''.join(str(x) for x in args)}")

    def read(self) -> str:
        output = self.inputs.pop()
        self.log.append(f"> {output}")
        return output

    def write_file(self, data: iter) -> None:
        self.log.append(f"! {'|'.join(data)}")

    def read_file(self) -> iter:
        self.log.append("* READ FILE")
        return self.file_content

    def now(self) -> datetime.datetime:
        timestamp = self.timestamps.pop()
        self.log.append(f"@ {timestamp}")
        return timestamp


def test_approve():
    inputs = ['N']
    timestamps = []
    file_content = []
    io = TestableIO(inputs, timestamps, file_content)
    run_main_(io)
    reporter = GenericDiffReporterFactory().get_first_working()
    verify("\n".join(io.log), reporter)
