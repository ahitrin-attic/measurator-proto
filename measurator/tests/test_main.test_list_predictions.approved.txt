(READ FILE)

Current time is 2020-01-31 12:00.

Program output:
> Successful predictions:41%, not done yet:2

Program output:
> Add another prediction? Yes/*No*/List

User input:
> L

Program output:
> 2020-03-01 08:00: NSDPRD-13544: уложусь ли в оценку 13d? 90% - НЕТ

Program output:
> 2020-03-01 08:00: NSDPRD-13544: уложусь ли по затратам в 26d? 90% - ДА

File is written:
```
F,2019-11-13 11:10,2019-11-18 09:00,,Успею ли я починить все падающие автотесты в NSDPRD-13267 до конца недели? 90%-нет
F,2019-11-13 11:10,2019-11-18 09:00,,Успею ли я до конца недели начать работать над нагрузочным тестированием NSDPRJ-2313? 90%-да
F,2019-11-14 13:10,2019-12-03 07:00,,"Задача NSDPRD-13339 (правки сортировки деревьев) потребует затрат не более, чем на 1d 4h"
S,2019-11-14 13:10,2020-01-20 08:00,,"Задача NSDPRD-12928 (перенос метода для фильтрации из бина в API) потребует затрат не более, чем на 5d"
S,2019-11-17 18:10,2019-11-18 07:00,,Упадут ли хоть какие-нибудь тесты на импорт в ветке NSDPRD-13267-pre1? 90% - да
S,2019-11-18 10:32,2019-11-23 00:00,,Успею ли я отдать NSDPRD-13267 в ревью до конца недели? 90%-да
F,2019-11-19 14:10,2019-11-26 07:00,,NSDPRD-13445: смогу ли уложиться в оценку трат на разработку 8h? 90% - да
S,2019-11-25 14:50,2019-11-30 07:00,,NSDPRD-13339: успею ли завершить в течение недели (без учёта церемоний)? 90% - да
F,2019-11-29 11:49,2019-12-04 08:00,,NSDPRD-13440: успею ли уложиться в оценку разработки 2d? 90% -да
S,2019-12-02 12:58,2019-12-07 08:00,,Успею ли взяться за задачу по Junit5 или PMD до конца недели? 90% - да
F,2019-12-05 12:58,2019-12-10 07:00,,NSDPRD-13575: уложусь ли в оценку 4h на работу? 90%-да
F,2019-12-09 13:19,2020-01-10 08:00,,NSDPRD-13547 (запуск PMD): уложусь ли в оценку затрат 12h?
N,2020-01-14 11:02,2020-03-01 08:00,,NSDPRD-13544: уложусь ли в оценку 13d? 90% - НЕТ
N,2020-01-14 11:02,2020-03-01 08:00,,NSDPRD-13544: уложусь ли по затратам в 26d? 90% - ДА
```